File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<pair<int, char>> st;
        int n = s.length();
       for(int i=0;i<n;i++) {
           if(s[i]==')' || s[i]=='(') {
               if(st.empty()) {
                   st.push({i, s[i]});
               }
               else if(s[i]==')' && st.top().second=='(') {
                   st.pop();
               }
               else {
                   st.push({i, s[i]});
               }
           }
       }
        set<int> hs;
        while(!st.empty()) {
            int index = st.top().first;
            hs.insert(index);
            st.pop();
        }
        string ans = "";
        for(int i=0;i<n;i++) {
            if(hs.find(i)==hs.end())
                ans += s[i];
        }
        return ans;
    } 
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        
      Stack<Pair<Character, Integer>> st = new Stack<>();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c == ')' || c == '('){
                if(st.empty()){
                    st.push(new Pair<>(c, i));
                } else{
                    if(c == ')' && st.peek().getKey() == '('){
                        st.pop();
                    } else{
                        st.push(new Pair<>(c, i));
                    }
                }
            }
        }

        Set<Integer> indexesToBeRemoved = new HashSet<>();
        while(!st.empty()){
            indexesToBeRemoved.add(st.peek().getValue());
            st.pop();
        }

        StringBuilder ans = new StringBuilder();
        for(int i=0;i<s.length();i++){
            if(!indexesToBeRemoved.contains(i)){
                ans.append(s.charAt(i));
            }
        }

        return ans.toString(); 
    }
}